llvm/lib/Target/AIE/AIEClusterBaseAddress.cpp

```diff
@@ -136,242 +117,152 @@ class AIEClusterBaseAddress : public MachineFunctionPass {
 
   StringRef getPassName() const override { return AIE_CLUSTER_BASE_ADDRESS; }
 
+  using RegToPtrAddMap = std::map<Register, SmallVector<MachineInstr *, 8>>;
+
 private:
-  bool processBasicBlock(MachineBasicBlock &MBB, MachineRegisterInfo &MRI,
-                         MachineIRBuilder &MIB,
+  MachineRegisterInfo *MRI = nullptr;
+  MachineDominatorTree *DT = nullptr;
+
+  bool processBasicBlock(MachineBasicBlock &MBB, MachineIRBuilder &MIB,
                          GISelObserverWrapper &Observer) {
-    /* Pass 1 :
-      Traverse MBB to try and connect G_LOAD or G_STORE to following PTR_ADDs.
-      This also checks if any MI uses the base ptr and if it does, we simply
-      remove the entire entry from the ChainedPtrAdds map. We avoid combining if
-      the base register of the to-be-chained ptradd is used by another instr.
-      This would otherwise generate a COPY and increase reg pressure.
-      Use a map to store the Load/Store MI and the next ptr add with the base
-      ptr of the Ld/St MI and the new updated offset.
-    */
-    std::map<MachineInstr *, PtrAdd> ChainedPtrAdds =
-        findChainablePtrAdds(MBB, MRI);
-
-    // Return false if ChainedPtrAdds is empty since we have no ptradds to
-    // update.
-    if (ChainedPtrAdds.empty())
-      return false;
-
-    /* Pass 2 :
-      Simply update the chainable ptradds in the MBB using the information in
-      the ChainedPtrAdds map.
-    */
-    updatePtrAddsInMBB(MBB, MRI, MIB, Observer, ChainedPtrAdds);
-
-    return true;
-  }
 
-  std::map<MachineInstr *, PtrAdd>
-  findChainablePtrAdds(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
-    std::map<MachineInstr *, PtrAdd> ChainedPtrAdds;
-    for (MachineInstr &MI : MBB) {
-      if (dyn_cast<GLoadStore>(&MI)) {
-        processLoadOrStore(MI, MRI, ChainedPtrAdds);
-        continue;
-      }
+    RegToPtrAddMap RegPtrUseMap;
+    bool Changed = false;
+
+    // Get all G_PTR_ADDs that use the same pointer.
+    getChainingCandidates(MBB, RegPtrUseMap);
```

Comment:
Super nit: I would rename this to e.g. `collectPtrUses()`, because we just collect registers used in ptr_adds, there is not extra processing.